glommio/src/read_write_lock.rs

```diff
@@ -0,0 +1,1214 @@
+    }
+    fn wake_up_readers_and_first_writer(rw: &mut State<T>) {
+        loop {
+            let waiter_id = rw.waiters.pop_front();
+            if let Some(waiter_id) = waiter_id {
+                let (waiter_kind, waker) = rw.waiters_map.remove(&waiter_id).unwrap();
+                waker.wake();
+
+                if waiter_kind == WaiterKind::WRITER {
+                    rw.queued_writers -= 1;
+                    break;
+                }
+            } else {
+                break;
+            }
+        }
+    }
+
+    fn wake_up_all_fibers(rw: &mut State<T>) {
+        for (_, (_, waker)) in rw.waiters_map.drain() {
+            waker.wake();
+        }
+
+        rw.waiters.clear();
+    }
+}
+
+impl<T> Drop for ReadWriteLock<T> {
+    fn drop(&mut self) {
+        self.close();
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::read_write_lock::LockClosedError;
+    use crate::read_write_lock::ReadWriteLock;
+    use crate::read_write_lock::TryReadWriteLockError;
+
+    use crate::Semaphore;
+    use rand::{self, Rng};
+    use std::cell::RefCell;
+    use std::rc::Rc;
+
+    #[derive(Eq, PartialEq, Debug)]
+    struct NonCopy(i32);
+
+    #[test]
+    fn test_smoke() {
+        test_executor!(async move {
+            let lock = ReadWriteLock::new(());
+            drop(lock.read().await.unwrap());
+            drop(lock.write().await.unwrap());
+            drop((lock.read().await.unwrap(), lock.read().await.unwrap()));
+            drop(lock.read().await.unwrap());
+        });
+    }
+
+    #[test]
+    fn test_frob() {
+        test_executor!(async move {
+            const N: u64 = 10;
+            const M: usize = 1000;
+
+            let r = Rc::new(ReadWriteLock::new(()));
+            let mut futures = Vec::new();
+
+            for _ in 0..N {
+                let r = r.clone();
+
+                let f = Task::local(async move {
+                    let mut rng = rand::thread_rng();
+                    for _ in 0..M {
+                        if rng.gen_bool(1.0 / (N as f64)) {
+                            drop(r.write().await.unwrap());
+                        } else {
+                            drop(r.read().await.unwrap());
+                        }
+                    }
+                });
+
+                futures.push(f);
+            }
+
+            join_all(futures).await;
+        });
+    }
+
+    #[test]
+    fn test_close_wr() {
+        test_executor!(async move {
+            let rc = Rc::new(ReadWriteLock::new(1));
+            let rc2 = rc.clone();
+
+            Task::local(async move {
+                let _lock = rc2.write().await.unwrap();
+                rc2.close();
+            })
+            .await;
+
+            assert!(rc.read().await.is_err());
+        });
+    }
+
+    #[test]
+    fn test_close_ww() {
+        test_executor!(async move {
+            let rc = Rc::new(ReadWriteLock::new(1));
+            let rc2 = rc.clone();
+
+            Task::local(async move {
```

Comment:
same issue as above, although this one could be solved in a different manner.
(one of them should fail, doesn't matter which)

However it is good to have the same logic everywhere. We can come up with a general test macro for this if it helps.

Reply:
Hi, the same as above I suppose that order of tasks is predefined because we are awaiting its execution to be finished. As I understand if we call `.dispatch()` at the end of task creation, the task will be executed at some moment in background, but if we call `.await` it is the same as we would call `.join()` once we spawn a new thread. Am I mistaken?